src/wallet/crypto.go

```diff
@@ -0,0 +1,405 @@
+var cryptoTable = map[CryptoType]cryptor{
+	CryptoTypeSha256Xor:              &sha256xorCrypto{},
+	CryptoTypeScryptChacha20poly1305: newScryptChacha20poly1305Crypto(scryptN, scryptR, scryptP, scryptKeyLen, scryptSaltLen),
+}
+
+// ErrAuthenticationFailed wraps the error of decryption.
+type ErrAuthenticationFailed struct {
+	err error
+}
+
+func (e ErrAuthenticationFailed) Error() string {
+	return e.err.Error()
+}
+
+// getCrypto gets crypto of given type
+func getCrypto(cryptoType CryptoType) (cryptor, error) {
+	c, ok := cryptoTable[cryptoType]
+	if !ok {
+		return nil, fmt.Errorf("can not find crypto %v in crypto table", cryptoType)
+	}
+
+	return c, nil
+}
+
+// scryptParams records the scrypt paramenters
+type scryptParams struct {
+	N       int    `json:"N"`       // scrypt N parament
+	R       int    `json:"r"`       // scrypt r parament
+	P       int    `json:"p"`       // scrypt p parament
+	KeyLen  int    `json:"key_len"` // scrypt keyLen parament
+	SaltLen int    `json:"-"`       // scrypt salt length
+	Salt    []byte `json:"salt"`    // scrypt salt parament
+}
+
+// chacha20poly1305Param records the chacha20poly1305 paramenters
+type chacha20poly1305Params struct {
+	Nonce          []byte `json:"nonce"`           // chacha20poly1305 nonce parament
+	AdditionalData []byte `json:"additional_data"` // chacha20poly1305 additionalData parament
+}
+
+// authenticated records the scrypt and chacha20poly1305 paramenters
+type authenticated struct {
+	scryptParams
+	chacha20poly1305Params
+}
+
+func (a *authenticated) Serialize() ([]byte, error) {
+	d, err := json.Marshal(a)
+	if err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+func (a *authenticated) Deserialize(d []byte) error {
+	if a == nil {
+		return errors.New("authenticated is nil")
+	}
+
+	return json.Unmarshal(d, a)
+}
+
+// scryptChacha20poly1305Crypto use scrypt and chacha20poly1305 to encrypt/decrypt wallet
+type scryptChacha20poly1305Crypto struct {
+	scryptParams // scrypt paramenters
+}
+
+func newScryptChacha20poly1305Crypto(N, r, p, keyLen, saltLen int) *scryptChacha20poly1305Crypto {
+	return &scryptChacha20poly1305Crypto{
+		scryptParams: scryptParams{
+			N:       N,
+			R:       r,
+			P:       p,
+			KeyLen:  keyLen,
+			SaltLen: saltLen,
+		},
+	}
+}
+
+// Encrypt encrypts wallet using scrypt and chacha20poly1305 with given password
+func (s *scryptChacha20poly1305Crypto) Encrypt(w *Wallet, password []byte) error {
+	if len(password) == 0 {
+		return ErrMissingPassword
+	}
+
+	if w.IsEncrypted() {
+		return ErrWalletEncrypted
+	}
+
+	wlt := w.clone()
+
+	// Derives password key
+	salt := cipher.RandByte(s.SaltLen)
+	dk, err := scrypt.Key(password, salt, s.N, s.R, s.P, s.KeyLen)
+	if err != nil {
+		return err
+	}
+
+	aead, err := chacha20poly1305.New(dk)
+	if err != nil {
+		return err
+	}
+
+	nonce := cipher.RandByte(chacha20poly1305.NonceSize)
+	// Generates additional data
+	ad := cipher.RandByte(additionalDataLen)
```

Comment:
what is random additional data for? the scrypt params need to go in here, random data is useless to put there. the point is that the additional data is also authenticated (but not encrypted)